File: examples/orders-service-py/orders/api/order_routes.py
"""Translate incoming HTTP requests into order-service calls."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from ..domain.order import Order, OrderRequest
from ..services.order_service import OrderService
from ..validation.order_validator import ValidationError


@dataclass
class ApiResponse:
    """A minimal HTTP-style response."""

    status: int
    body: Any


class OrderRoutes:
    """The system's front door."""

    def __init__(self, orders: OrderService) -> None:
        self._orders = orders

    def handle_create_order(self, request: OrderRequest) -> ApiResponse:
        """POST /orders — place a new order."""
        try:
            order = self._orders.place_order(request)
            return self._created(order)
        except ValidationError as error:
            return ApiResponse(status=400, body={"error": str(error)})

    def handle_get_order(self, order_id: str) -> ApiResponse:
        """GET /orders/:id — fetch a previously placed order."""
        order = self._orders.get_order(order_id)
        if order is None:
            return ApiResponse(status=404, body={"error": "order not found"})
        return ApiResponse(status=200, body=order)

    def _created(self, order: Order) -> ApiResponse:
        """Shape a 201 response for a freshly placed order."""
        return ApiResponse(status=201, body=order)

File: examples/orders-service/src/api/orderRoutes.ts
import type { Order, OrderRequest } from "../domain/order.js";
import { OrderService } from "../services/orderService.js";
import { ValidationError } from "../validation/orderValidator.js";

/** A minimal HTTP-style response. */
export interface ApiResponse {
  status: number;
  body: unknown;
}

/** Translates incoming HTTP requests into order-service calls. The system's front door. */
export class OrderRoutes {
  constructor(private readonly orders: OrderService) {}

  /** POST /orders — place a new order. */
  handleCreateOrder(request: OrderRequest): ApiResponse {
    try {
      const order = this.orders.placeOrder(request);
      return this.created(order);
    } catch (error) {
      return this.toErrorResponse(error);
    }
  }

  /** GET /orders/:id — fetch a previously placed order. */
  handleGetOrder(id: string): ApiResponse {
    const order = this.orders.getOrder(id);
    if (!order) {
      return { status: 404, body: { error: "order not found" } };
    }
    return { status: 200, body: order };
  }

  /** Shape a 201 response for a freshly placed order. */
  private created(order: Order): ApiResponse {
    return { status: 201, body: order };
  }

  /** Map domain errors onto HTTP status codes. */
  private toErrorResponse(error: unknown): ApiResponse {
    if (error instanceof ValidationError) {
      return { status: 400, body: { error: error.message } };
    }
    return { status: 500, body: { error: "internal error" } };
  }
}
